internal/resource/url.go

```diff
@@ -522,3 +545,41 @@ func (f *Fetcher) decompressCopyHashAndVerify(dest io.Writer, src io.Reader, opt
 	}
 	return nil
 }
+
+// parseARN is a custom wrapper around arn.Parse(); it takes arnURL, a full ARN URL,
+// and returns a bucket, a key, a potentially empty region, or an error if the ARN
+// is invalid or not for an S3 object.
+// If the given arnURL is an accesspoint ARN, the region is set.
+// The region is empty for S3 bucket ARNs because they don't include the region field.
+func (f *Fetcher) parseARN(arnURL string) (string, string, string, error) {
+	if !arn.IsARN(arnURL) {
+		return "", "", "", configErrors.ErrInvalidS3ARN
+	}
+	s3arn, err := arn.Parse(arnURL)
+	if err != nil {
+		return "", "", "", err
+	}
+	if s3arn.Service != "s3" {
+		return "", "", "", configErrors.ErrInvalidS3ARN
+	}
+
+	// Determine if the ARN is for an access point or a bucket.
+	var bucket, key string
+	urlSplit := strings.Split(arnURL, "/")
```

Comment:
The code below this uses indexing into this slice directly; are we 100% sure that no URL could get to here (i.e. past `arn.IsARN()` and `arn.Parse()`) such that those indexes could never cause a panic?

Would it be safer to just do sanity-checks of its length in each case (accesspoint vs not) and nicely return an error if it's not the minimum we expect?

Reply:
Sure, I added some sanity length checks to this function and in the URL validation, and added test cases to check this. Unfortunately the ARN package just handles splitting the ARN parts delimited by `:`. Since resources could be from other services (ex. IAM), they don't expose a way to index or split the resource part. I added the relevant documentation https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html#arns-paths in a comment there as well.